File: go/pkg/basecamp/client_replies.go
package basecamp

import (
	"context"
	"fmt"
	"time"

	"github.com/basecamp/basecamp-sdk/go/pkg/generated"
)

// ClientReply represents a reply to a client correspondence or approval.
type ClientReply struct {
	ID               int64     `json:"id"`
	Status           string    `json:"status"`
	VisibleToClients bool      `json:"visible_to_clients"`
	CreatedAt        time.Time `json:"created_at"`
	UpdatedAt        time.Time `json:"updated_at"`
	Title            string    `json:"title"`
	InheritsStatus   bool      `json:"inherits_status"`
	Type             string    `json:"type"`
	URL              string    `json:"url"`
	AppURL           string    `json:"app_url"`
	BookmarkURL      string    `json:"bookmark_url"`
	Parent           *Parent   `json:"parent,omitempty"`
	Bucket           *Bucket   `json:"bucket,omitempty"`
	Creator          *Person   `json:"creator,omitempty"`
	Content          string    `json:"content"`
}

// ClientRepliesService handles client reply operations.
type ClientRepliesService struct {
	client *AccountClient
}

// NewClientRepliesService creates a new ClientRepliesService.
func NewClientRepliesService(client *AccountClient) *ClientRepliesService {
	return &ClientRepliesService{client: client}
}

// List returns all replies for a client recording (correspondence or approval).
// bucketID is the project ID, recordingID is the parent correspondence/approval ID.
func (s *ClientRepliesService) List(ctx context.Context, bucketID, recordingID int64) (result []ClientReply, err error) {
	op := OperationInfo{
		Service: "ClientReplies", Operation: "List",
		ResourceType: "client_reply", IsMutation: false,
		BucketID: bucketID, ResourceID: recordingID,
	}
	if gater, ok := s.client.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.parent.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	resp, err := s.client.gen.ListClientRepliesWithResponse(ctx, s.client.accountID, bucketID, recordingID)
	if err != nil {
		return nil, err
	}
	if err = checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, nil
	}

	replies := make([]ClientReply, 0, len(*resp.JSON200))
	for _, gr := range *resp.JSON200 {
		replies = append(replies, clientReplyFromGenerated(gr))
	}

	return replies, nil
}

// Get returns a specific client reply.
// bucketID is the project ID, recordingID is the parent correspondence/approval ID,
// replyID is the client reply ID.
func (s *ClientRepliesService) Get(ctx context.Context, bucketID, recordingID, replyID int64) (result *ClientReply, err error) {
	op := OperationInfo{
		Service: "ClientReplies", Operation: "Get",
		ResourceType: "client_reply", IsMutation: false,
		BucketID: bucketID, ResourceID: replyID,
	}
	if gater, ok := s.client.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.parent.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	resp, err := s.client.gen.GetClientReplyWithResponse(ctx, s.client.accountID, bucketID, recordingID, replyID)
	if err != nil {
		return nil, err
	}
	if err = checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		err = fmt.Errorf("unexpected empty response")
		return nil, err
	}

	reply := clientReplyFromGenerated(resp.JSON200.Reply)
	return &reply, nil
}

// clientReplyFromGenerated converts a generated ClientReply to our clean type.
func clientReplyFromGenerated(gr generated.ClientReply) ClientReply {
	r := ClientReply{
		Status:           gr.Status,
		VisibleToClients: gr.VisibleToClients,
		CreatedAt:        gr.CreatedAt,
		UpdatedAt:        gr.UpdatedAt,
		Title:            gr.Title,
		InheritsStatus:   gr.InheritsStatus,
		Type:             gr.Type,
		URL:              gr.Url,
		AppURL:           gr.AppUrl,
		BookmarkURL:      gr.BookmarkUrl,
		Content:          gr.Content,
	}

	if gr.Id != nil {
		r.ID = *gr.Id
	}

	if gr.Parent.Id != nil || gr.Parent.Title != "" {
		r.Parent = &Parent{
			ID:     derefInt64(gr.Parent.Id),
			Title:  gr.Parent.Title,
			Type:   gr.Parent.Type,
			URL:    gr.Parent.Url,
			AppURL: gr.Parent.AppUrl,
		}
	}

	if gr.Bucket.Id != nil || gr.Bucket.Name != "" {
		r.Bucket = &Bucket{
			ID:   derefInt64(gr.Bucket.Id),
			Name: gr.Bucket.Name,
			Type: gr.Bucket.Type,
		}
	}

	if gr.Creator.Id != nil || gr.Creator.Name != "" {
		r.Creator = &Person{
			ID:           derefInt64(gr.Creator.Id),
			Name:         gr.Creator.Name,
			EmailAddress: gr.Creator.EmailAddress,
			AvatarURL:    gr.Creator.AvatarUrl,
			Admin:        gr.Creator.Admin,
			Owner:        gr.Creator.Owner,
		}
	}

	return r
}

File: typescript/src/services/client-replies.ts
/**
 * Client Replies service for the Basecamp API.
 *
 * Client replies are responses to client correspondences or approvals
 * within a project's client portal.
 *
 * @example
 * ```ts
 * const replies = await client.clientReplies.list(projectId, recordingId);
 * const reply = await client.clientReplies.get(projectId, recordingId, replyId);
 * ```
 */

import { BaseService } from "./base.js";

// =============================================================================
// Types
// =============================================================================

/**
 * A person reference (simplified).
 */
export interface PersonRef {
  id: number;
  name: string;
  email_address?: string;
  avatar_url?: string;
  admin?: boolean;
  owner?: boolean;
}

/**
 * A bucket (project) reference.
 */
export interface BucketRef {
  id: number;
  name: string;
  type: string;
}

/**
 * A parent reference.
 */
export interface ParentRef {
  id: number;
  title: string;
  type: string;
  url: string;
  app_url: string;
}

/**
 * A reply to a client correspondence or approval.
 */
export interface ClientReply {
  id: number;
  status: string;
  visible_to_clients: boolean;
  created_at: string;
  updated_at: string;
  title: string;
  inherits_status: boolean;
  type: string;
  url: string;
  app_url: string;
  bookmark_url: string;
  content: string;
  parent?: ParentRef;
  bucket?: BucketRef;
  creator?: PersonRef;
}

// =============================================================================
// Service
// =============================================================================

/**
 * Service for managing client replies in Basecamp.
 */
export class ClientRepliesService extends BaseService {
  /**
   * Lists all replies for a client recording (correspondence or approval).
   *
   * @param projectId - The project (bucket) ID
   * @param recordingId - The parent correspondence/approval ID
   * @returns Array of client replies
   *
   * @example
   * ```ts
   * const replies = await client.clientReplies.list(projectId, correspondenceId);
   * replies.forEach(r => console.log(r.creator?.name, r.content));
   * ```
   */
  async list(projectId: number, recordingId: number): Promise<ClientReply[]> {
    const response = await this.request(
      {
        service: "ClientReplies",
        operation: "List",
        resourceType: "client_reply",
        isMutation: false,
        projectId,
        resourceId: recordingId,
      },
      () =>
        this.client.GET("/buckets/{projectId}/client/recordings/{recordingId}/replies.json", {
          params: { path: { projectId, recordingId } },
        })
    );

    return (response ?? []) as ClientReply[];
  }

  /**
   * Gets a specific client reply.
   *
   * @param projectId - The project (bucket) ID
   * @param recordingId - The parent correspondence/approval ID
   * @param replyId - The client reply ID
   * @returns The client reply
   * @throws BasecampError with code "not_found" if reply doesn't exist
   *
   * @example
   * ```ts
   * const reply = await client.clientReplies.get(projectId, recordingId, replyId);
   * console.log(reply.content, reply.creator?.name);
   * ```
   */
  async get(projectId: number, recordingId: number, replyId: number): Promise<ClientReply> {
    const response = await this.request(
      {
        service: "ClientReplies",
        operation: "Get",
        resourceType: "client_reply",
        isMutation: false,
        projectId,
        resourceId: replyId,
      },
      () =>
        this.client.GET("/buckets/{projectId}/client/recordings/{recordingId}/replies/{replyId}", {
          params: { path: { projectId, recordingId, replyId } },
        })
    );

    return response as unknown as ClientReply;
  }
}
